File: graph/js/Graph.js
class Graph {
  constructor() {
    this.graph = {};
    this.visited = {};
    this.parents = {};
    this.stack = [];
  }

  add(a, b, isDirected = false) {
    // Always make sure nodes exist in graph
    if (!this.graph[a]) this.graph[a] = [];
    if (!this.graph[b]) this.graph[b] = [];

    // Initialize visited & parents for new nodes
    if (!(a in this.visited)) this.visited[a] = false;
    if (!(b in this.visited)) this.visited[b] = false;
    if (!(a in this.parents)) this.parents[a] = null;
    if (!(b in this.parents)) this.parents[b] = null;

    // Add the edge
    this.graph[a].push(b);
    if (!isDirected) {
      this.graph[b].push(a);
    }
  }

  BFSTraverse(src) {
    let queue = [src];
    this.visited[src] = true;

    if (!this.graph[src]) {
      console.log("Data not exist");
      return;
    }

    while (queue.length != 0) {
      const value = queue.shift();
      console.log(value);

      let arr = this.graph[value];

      for (let i = 0; i < arr.length; i++) {
        if (!this.visited[arr[i]]) {
          queue.push(arr[i]);
          this.visited[arr[i]] = true;
        }
      }
    }

    this.resetVisited();
  }

  DFSTraverse(src) {
    this.visited[src] = true;
    console.log(src);
    let arr = this.graph[src];
    for (let i = 0; i < arr.length; i++) {
      if (!this.visited[arr[i]]) {
        this.DFSTraverse(arr[i]);
      }
    }
  }

  resetVisited() {
    for (let node in this.visited) {
      this.visited[node] = false;
    }
  }

  resetParents() {
    for (let node in this.parents) {
      this.parents[node] = null;
    }
  }

  cycleDetectionBFS(src) {
    this.visited[src] = true;
    this.parents[src] = null;
    const queue = [src];

    while (queue.length != 0) {
      const value = queue.shift();
      const arr = this.graph[value];

      for (let index = 0; index < arr.length; index++) {
        if (!this.visited[arr[index]]) {
          this.visited[arr[index]] = true;
          this.parents[arr[index]] = value;
          queue.push(arr[index]);
        } else if (this.parents[value] != arr[index]) {
          return "cycle detect";
        }
      }
    }

    this.resetParents();
    this.resetVisited();

    return "can't detect cycle";
  }

  shortestPathAlgoBFS(source, destination) {
    const queue = [source];
    this.visited[source] = true;
    this.parents[source] = null;

    while (queue.length != 0) {
      const value = queue.shift();
      const arr = this.graph[value];

      for (let index = 0; index < arr.length; index++) {
        if (!this.visited[arr[index]]) {
          queue.push(arr[index]);
          this.visited[arr[index]] = true;
          this.parents[arr[index]] = value;
        }
      }
    }

    // Reconstruct path
    const path = [destination];
    while (path[path.length - 1] !== source) {
      const parent = this.parents[path[path.length - 1]];
      if (parent === null || parent === undefined) {
        console.log("No path found.");
        return;
      }
      path.push(parent);
    }

    path.reverse();
    console.log("Path:", path.join(" -> "));
    this.resetVisited();
    this.resetParents();
  }

  bipartiteAlgoBFS(src, color1, color2) {
    const queue = [src];
    const color = {};
    this.visited[src] = true;
    color[src] = color1;

    while (queue.length != 0) {
      const value = queue.shift();
      const arr = this.graph[value];

      for (let node of arr) {
        if (!this.visited[node]) {
          color[node] = color[value] === color1 ? color2 : color1;
          this.visited[node] = true;
          queue.push(node);
        } else if (color[value] === color[node]) {
          this.resetVisited();
          return "Graph is NOT bipartite";
        }
      }
    }
    this.resetVisited();
    return "Graph is  bipartite";
  }

  topologySortingDFS(src) {
    this.visited[src] = true;
    let arr = this.graph[src];
    
    for (let i = 0; i < arr.length; i++) {
      if (!this.visited[arr[i]]) {
        this.topologySortingDFS(arr[i]);
      }
    }

    this.stack.push(src);
  }
}

const gp = new Graph();
gp.add(0, 1, true);
gp.add(1, 2, true);
gp.add(3, 2, true);
gp.add(2, 4, true);
gp.add(4, 5, true);
gp.add(6, 5, true);
// gp.add(6, 3);
// gp.BFSTraverse(3);
// gp.DFSTraverse(3);
// gp.resetVisited();
// console.log(gp.cycleDetectionBFS(0));
// console.log(gp)
// gp.shortestPathAlgoBFS(6, 5);
// console.log(gp.bipartiteAlgoBFS(0, "red", "green"));

for (const node in gp.visited) {
  if (!gp.visited[node]) {
    gp.topologySortingDFS(node);
  }
}

console.log(gp.stack.reverse().join("->"));
gp.resetVisited();

File: graph/py/Graph.py

class Graph:
    def __init__(self):
        self.graph = {}
        self.visited = {}
        self.parents = {}
        self.stack = []

    def add_gp(self, a, b, is_directed=False):
        # Ensure both nodes exist in the graph dictionary
        if a not in self.graph:
            self.graph[a] = []
        if b not in self.graph:
            self.graph[b] = []

       # Initialize visited and parents
        if a not in self.visited:
           self.visited[a] = False
        if b not in self.visited:
           self.visited[b] = False
        if a not in self.parents:
           self.parents[a] = None
        if b not in self.parents:
           self.parents[b] = None

        # Add the edge
        self.graph[a].append(b)
        if not is_directed:
           self.graph[b].append(a)

    def traverse_BFS(self,src):
     self.visited[src] = True
     queue = [src]

     while queue:
      val = queue.pop(0)
      print(val)

      for neighbor in self.graph[val]:
         if not self.visited[neighbor]:
          self.visited[neighbor] = True
          queue.append(neighbor)

     self.reset_visited()

    def traverse_DFS(self,src):
       
       self.visited[src] = True
       print(src)

       for neighbor in self.graph[src]:
         if not self.visited[neighbor]:
           self.traverse_DFS(neighbor)

    def cycle_detection(self,src):
        self.visited[src] = True
        self.parents[src] = None
        queue = [src]

        while queue:
          val = queue.pop(0)

          for neighbor in self.graph[val]:
             if not self.visited[neighbor]:
               self.visited[neighbor] = True
               self.parents[neighbor] = val
               queue.append(neighbor)
             elif(self.parents[val] != neighbor):
               return "cycle detection"
             
        self.reset_parents()
        self.reset_visited()
             
        return "not found cycle"
    
    def shortest_path_BFS(self,source,destination):
     self.visited[source] = True
     self.parents[source] = None
     queue = [source]

     while queue:
       val = queue.pop(0)

       for neighbor in self.graph[val]:
          if not self.visited[neighbor]:
            self.visited[neighbor] = True
            self.parents[neighbor] = val
            queue.append(neighbor)

     path = [destination]

     while True:
      path.append(self.parents[path[-1]])
      if source == path[-1]:
        break
     
     self.reset_parents()
     self.reset_visited()

     path.reverse()
     actual_path = ">".join(map(str,path))
     return actual_path

    def reset_parents(self):
     for item in self.parents:
        self.parents[item] = None

    def reset_visited(self):
     for node in self.visited:
        self.visited[node] = False

    def bipartite_algo_BFS(self,src,color1,color2):
       self.visited[src] = True
       color = {}
       color[src] = color1
       queue = [src]


       while queue:
         val = queue.pop(0)

         for neighbor in self.graph[val]:
            if not self.visited[neighbor]:
              color[neighbor] = color1 if color[val] == color2 else color2
              self.visited[neighbor] = True
              queue.append(neighbor)
            elif color[val] == color[neighbor]:
              print("Graph is NOT bipartite")
              return

       self.reset_visited()

       print("Graph Is  bipartite")
       return
    
    def topology_sorting_DFS(self,src):
     self.visited[src] = True

     for neighbor in self.graph[src]:
       if not self.visited[neighbor]:
         self.topology_sorting_DFS(neighbor)

     self.stack.append(src)
      


gp = Graph()
gp.add_gp(0,1,True)
gp.add_gp(1,2,True)
gp.add_gp(1,7,True)
gp.add_gp(2,3,True)
gp.add_gp(3,4,True)
gp.add_gp(4,5,True)
gp.add_gp(5,6,True)
gp.add_gp(6,7,True)


print(gp.graph)
gp.traverse_BFS(0)
# gp.traverse_DFS(0)
# print(gp.visited)
print(gp.cycle_detection(0))
print(gp.shortest_path_BFS(1,4))
gp.bipartite_algo_BFS(0,"red","green")
# print(gp.parents)
gp.reset_parents()
gp.reset_visited()

for node in gp.visited:
  if not gp.visited[node]:
    gp.topology_sorting_DFS(node)

gp.stack.reverse()
actual_topology_sorting_path = ">".join(map(str,gp.stack))
print(actual_topology_sorting_path)
 

gp.reset_parents()
gp.reset_visited()
gp.traverse_DFS(0)


   

    
        
        

       
